### PR TITLE
[VS] Scripting fixes

### PR DESCRIPTION
**Pull request by TIHan**
We have a few issues with F# scripts getting re-analyzed when their dependencies have been updated.

The idea here is to create a VS project that tracks the dependencies of the script so Roslyn can accurately re-analyze it.

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs [hunk 1]
```diff
@@ -31,36 +33,91 @@ type internal SingleFileWorkspaceMap(workspace: VisualStudioWorkspace,
...
-                | true, projectContext ->
-                    optionsManager.ClearSingleFileOptionsCache(document.Id)
-                    projectContext.Dispose()
-                | _ -> ()
+                optionsManager.ClearSingleFileOptionsCache(document.Id)
+
+                lock gate (fun () ->
+                    match files.TryRemove(document.FilePath) with
+                    | true, (projectContext, _) ->
+                        projectContext.Dispose()
```
Comment (TIHan) [inline vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs:118]: We should not dispose of a `projectContext` if it's being used as a dependency for another project.

**Comment by TIHan** [comment]
It would be preferred we finish https://github.com/dotnet/fsharp/pull/11588 first as it will tie into how scripts will be updated.

**Comment by dsyme** [comment]
So how do we get this under test? :)

**Review by dsyme — APPROVED** [review]

File: vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs (hunk 1 above)
Comment (dsyme) [reply vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs:118]: Note this TODO

**Comment by dsyme** [comment]
@TIHan Can we get this in now the refactoring PR is done?

**Comment by TIHan** [comment]
I'll see if we can now. I need to think about this as it might change because of the in-memory-documents PR.

File: vsintegration/tests/UnitTests/Workspace/WorkspaceTests.fs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    [<Test>]
+    let ``Script file referencing another script will correct update when the referenced script file changes``() =
+        let workspace = createWorkspace()
+        let miscFilesWorkspace = createMiscFileWorkspace()
+        let projectContextFactory = TestFSharpWorkspaceProjectContextFactory(workspace, miscFilesWorkspace)
+
+        let _miscFileService = FSharpMiscellaneousFileService(workspace, miscFilesWorkspace, projectContextFactory)
+
+        let filePath1 =
```
Comment (dsyme) [inline vsintegration/tests/UnitTests/Workspace/WorkspaceTests.fs:223]: Can be done later - but best to create an IDisposable thingy for these

**Comment by dsyme** [comment]
It would be good to get further tests added where we check that re-analysis happens when 

1. a `#r` referenecd DLL is added and removed+added

2. a `#r "nuget: .."` package is deleted and re-restored (do we expect to handle this?)

3. a type provider invalidation occurs

I'm particularly concerned about the last one as I don't currently understand how Roslyn could know that it needs to do a re-analysis in this case.

**Comment by dsyme** [comment]
I will merge this, as the fixes are important and the core testing added, and start using it myself. 

@TIHan Can you look at the above testing cases?

**Comment by TIHan** [comment]
Yes I will look at the test cases and get them in today.

**Pull request merged**